src/calories.rs

```diff
@@ -0,0 +1,92 @@
+#[derive(Debug, PartialEq)]
+pub struct CalorieCounter {
+    third_max: usize,
+    second_max: usize,
+    pub max: usize,
+    current_counter: usize,
+}
+
+impl CalorieCounter {
+    pub fn new(
+        third_max: usize,
+        second_max: usize,
+        max: usize,
+        current_counter: usize,
+    ) -> CalorieCounter {
+        CalorieCounter {
+            third_max,
+            second_max,
+            max,
+            current_counter,
+        }
+    }
+
+    pub fn reset() -> CalorieCounter {
+        CalorieCounter::new(0, 0, 0, 0)
+    }
+
+    pub fn with_next(&self, input: &str) -> CalorieCounter {
+        if input.is_empty() {
+            CalorieCounter::new(self.third_max, self.second_max, self.max, 0)
+        } else {
+            let input_int: usize = input.parse().unwrap();
+            let new_count = self.current_counter + input_int;
+            if new_count <= self.third_max {
+                CalorieCounter::new(self.third_max, self.second_max, self.max, new_count)
+            } else if new_count <= self.second_max {
+                CalorieCounter::new(new_count, self.second_max, self.max, new_count)
+            } else if new_count <= self.max {
+                CalorieCounter::new(self.second_max, new_count, self.max, new_count)
+            } else {
+                CalorieCounter::new(self.second_max, self.max, new_count, new_count)
+            }
+        }
+    }
+
+    pub fn fold_step(last_counter: CalorieCounter, new_input: String) -> CalorieCounter {
+        last_counter.with_next(&new_input)
```

Comment:
Arguably having this function wrapper around a method isn't adding much - could just put the body of `with_next()` directly in here.